pkg/controllers/jobflow/jobflow_controller_action.go

```diff
@@ -0,0 +1,309 @@
+		}
+	}
+
+	jobFlowStatus := v1alpha1flow.JobFlowStatus{
+		PendingJobs:    pendingJobs,
+		RunningJobs:    runningJobs,
+		FailedJobs:     FailedJobs,
+		CompletedJobs:  CompletedJobs,
+		TerminatedJobs: TerminatedJobs,
+		UnKnowJobs:     UnKnowJobs,
+		JobStatusList:  jobStatusList,
+		Conditions:     conditions,
+		State:          jobFlow.Status.State,
+	}
+	return &jobFlowStatus, nil
+}
+
+func getRunningHistories(jobStatusList []v1alpha1flow.JobStatus, job *v1alpha1.Job) []v1alpha1flow.JobRunningHistory {
+	runningHistories := make([]v1alpha1flow.JobRunningHistory, 0)
+	flag := true
+	for _, jobStatusGet := range jobStatusList {
+		if jobStatusGet.Name == job.Name {
+			if jobStatusGet.RunningHistories != nil {
+				flag = false
+				runningHistories = jobStatusGet.RunningHistories
+				// State change
+				if len(runningHistories) == 0 {
+					continue
+				}
+				if runningHistories[len(runningHistories)-1].State != job.Status.State.Phase {
+					runningHistories[len(runningHistories)-1].EndTimestamp = metav1.Time{
+						Time: time.Now(),
+					}
+					runningHistories = append(runningHistories, v1alpha1flow.JobRunningHistory{
+						StartTimestamp: metav1.Time{Time: time.Now()},
+						EndTimestamp:   metav1.Time{},
+						State:          job.Status.State.Phase,
+					})
+				}
+			}
+		}
+	}
+	if flag && job.Status.State.Phase != "" {
+		runningHistories = append(runningHistories, v1alpha1flow.JobRunningHistory{
+			StartTimestamp: metav1.Time{
+				Time: time.Now(),
+			},
+			EndTimestamp: metav1.Time{},
+			State:        job.Status.State.Phase,
+		})
+	}
+	return runningHistories
+}
+
+func (jf *jobflowcontroller) loadJobTemplateAndSetJob(jobFlow *v1alpha1flow.JobFlow, flowName string, jobName string, job *v1alpha1.Job) error {
+	// load jobTemplate
+	jobTemplate, err := jf.jobTemplateLister.JobTemplates(jobFlow.Namespace).Get(flowName)
+	if err != nil {
+		return err
+	}
+
+	*job = v1alpha1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        jobName,
+			Namespace:   jobFlow.Namespace,
+			Labels:      map[string]string{CreatedByJobTemplate: GetTemplateString(jobFlow.Namespace, flowName)},
+			Annotations: map[string]string{CreatedByJobTemplate: GetTemplateString(jobFlow.Namespace, flowName)},
+		},
+		Spec:   jobTemplate.Spec,
+		Status: v1alpha1.JobStatus{},
+	}
+
+	return controllerutil.SetControllerReference(jobFlow, job, scheme.Scheme)
+}
+
+func (jf *jobflowcontroller) deleteAllJobsCreateByJobFlow(jobFlow *v1alpha1flow.JobFlow) error {
+	selector := labels.NewSelector()
+	jobList, err := jf.jobLister.Jobs(jobFlow.Namespace).List(selector)
+	if err != nil {
+		return err
+	}
+
+	for _, job := range jobList {
+		if len(job.OwnerReferences) > 0 {
+			for _, reference := range job.OwnerReferences {
+				if reference.Kind == helpers.JobFlowKind.Kind && reference.Name == jobFlow.Name {
+					if err := jf.vcClient.BatchV1alpha1().Jobs(jobFlow.Namespace).Delete(context.Background(), job.Name, metav1.DeleteOptions{}); err != nil {
```

Comment:
Since this delete action code is a little bit long, suggest to judge the value of `err` at a new line.